### ISSUE TITLE
Type narrowing of generic type appears to have no effect

### ISSUE THREAD

**Issue by 0xjc** [microsoft/pyright#3237]
The following code fails type checking:

```python
# pyright: strict
from typing import TypeVar


class A:
    pass


T = TypeVar("T")


def f(a: T) -> T:
    if type(a) is A:
        return A()
    else:
        return a

```

The error:

```
  /tmp/testing.py:14:16 - error: Expression of type "A" cannot be assigned to return type "T@f"
    Type "A" cannot be assigned to type "T@f" (reportGeneralTypeIssues)
```

It seems that `type(a) is A` should narrow `a` to have the exact type `A` and therefore `T = A`. But the type checker believes they are not the same.

Version: pyright 1.1.231

**Comment by erictraut** [comment]
This current behavior is intended. Mypy produces the same error in this case.

Pyright will allow you to do the following, but mypy also produces an error in this case.
```python
        return type(a)()
```

That said, I think you are correct that the code sample is technically type safe and conforms to PEP 484. I'll think about it more.

**Comment by erictraut** [comment]
I think this is technically type safe, but implementing this is very difficult (probably not feasible) given pyright's architecture. The problem is that the return type `A()` has nothing to do with any type narrowing that takes place for the expression `a`. If the return type were based on the expression `a`, it would be possible. That's why `return type(a)()` type checks without error.

You'll need to find some other way to code this if you want it to type check with pyright (or mypy, for that matter).
